diff -ur qemu/target-arm/translate.c qemu-0.9.0/target-arm/translate.c
--- qemu/target-arm/translate.c	2007-02-06 00:01:54.000000000 +0100
+++ qemu-0.9.0/target-arm/translate.c	2007-07-10 09:15:56.000000000 +0200
@@ -2536,7 +2536,9 @@
                     int (*cpu_fprintf)(FILE *f, const char *fmt, ...),
                     int flags)
 {
+    uint32_t psr;
     int i;
+#ifndef X49GP
     union {
         uint32_t i;
         float s;
@@ -2548,7 +2550,7 @@
         float64 f64;
         double d;
     } d0;
-    uint32_t psr;
+#endif
 
     for(i=0;i<16;i++) {
         cpu_fprintf(f, "R%02d=%08x", i, env->regs[i]);
@@ -2567,6 +2567,7 @@
                 psr & CPSR_T ? 'T' : 'A', 
                 cpu_mode_names[psr & 0xf], (psr & 0x10) ? 32 : 26);
 
+#ifndef X49GP
     for (i = 0; i < 16; i++) {
         d.d = env->vfp.regs[i];
         s0.i = d.l.lower;
@@ -2579,5 +2580,6 @@
                     d0.d);
     }
     cpu_fprintf(f, "FPSCR: %08x\n", (int)env->vfp.xregs[ARM_VFP_FPSCR]);
+#endif
 }
 
